accepted/00080-api-versioning.md

```diff
@@ -0,0 +1,295 @@
+client applications or scripts compatible with multiple versions of
+Uynui and SUSE Manager (SUMA) API and suggest an improvement in this
+area.
+
+When the consumers use the API, there must be a way for them to make
+sure they call the endpoints correctly (they need to call an existing
+method with correct parameters and they also need to make some
+assumptions on the return value structure).
+
+TODO keep here? :
+The secondary goal is to provide better guidelines for writing scripts
+that consume the API in the `API > FAQ` section of the Uyuni/SUMA Web
+UI or in the documentation.
+
+
+## XMLRPC compatibility
+
+Uyuni/SUMA exposes the information about the API and its structure in
+3 places:
+
+1. Information about the versions via the `api` XMLRPC endpoint:
+   - `getVersion`: the API version, e.g. `25`. This integer grows with
+   time, but there are no strict rules about it.
+   - `systemVersion`: the Uyuni server version (e.g. `4.2.0 Beta1` for
+   SUMA, `2021.05` for Uyuni)
+
+2. Furthermore, the `api` namespace also exposes basic API
+   introspection calls describing the structure of the call:
+   - `getApiNamespaces`: all API namespaces
+   - `getApiNamespaceCallList`: API methods of given namespace
+     together with their parameter types, faults and return types
+   - `getApiCallList`: all API methods grouped by namespace
+
+3. Faults (retrospective): When the API consumers call a method in a
+   wrong way (incorrect parameters, missing method or namespace), they
+   get back an XMLRPC fault with code `-1`.
+   For instance:
+   ```python
+   # Non-existing method/wrong method signature
+   <Fault -1: 'redstone.xmlrpc.XmlRpcFault: Could not find method:
+   test in class: com.redhat.rhn.frontend.xmlrpc.ansible.AnsibleHandler with params: []'>
+
+   # Non-existing handler
+   <Fault -1: 'The specified handler cannot be found'>
+   ```
+   Note: This fault code is
+   not reserved solely to report the API calling mismatch cases, but
+   is used in other cases (see the`CustomInfoHandler.java` file in
+   Uyuni).
+
+## Existing approach and problems
+
+We must provide a sane way for users to write robust scripts that
+target various Uyuni/SUMA versions.
+
+In the `spacecmd` tool, this is done by API version check endpoint
+(`api.getVersion` above). The logic simply checks the API version and
+reacts accordingly (e.g. warns the user, that given operation is not
+supported by the API version).
+
+A similar approach is taken by the `errata-import.pl` script by Steve
+Meier
+[1](https://github.com/stevemeier/cefs/blob/master/errata-import.pl). The
+script contains a list of supported versions and checks on runtime,
+whether the API version of the Uyuni/SUMA server is contained in this
+list.
+
+This works well, until the user needs to target both Uyuni and SUSE
+Manager in their scripts - the API version numbers are not consistent
+(API version `25` in Uyuni is currently not the same as version `25`
+in SUMA).
```

Comment:
Currently 25 only exists in Uyuni. SUMA 4.2 will get 25 as soon as it is released. But Uyuni is changing (breaking and non breaking) but keep 25. This is one problem.
SUMA 4.1 is 24, but everytime when we backport a function to 4.1 we change the API but do not change the version.
Checking if a new funtion exists in 4.1 is not possible using the version

Reply:
Right. With this paragraph I only wanted to say `UYUNI API 25 != SUMA API 25`. I'll fix at least the fact that SUMA does not have `25` yet.